Potential PID file race condition causes `sr3 restart` to crash

I first noticed this about 2 weeks ago when I was running my auto-restart script on our stage servers. I wasn't able to catch the error from the stderr of my script output, but I was able to catch it today when I did a manual restart from the same stage server. 

What seems to happen:
1. `sr3 restart` gets invoked
2. At least one process is detected in a `missing` state. 
3. `_clean_missing_proc_state_dir` gets invoked to cleanup the directory of the missing process. Somehow, there's some kind of a race condition where the `*pid` file goes missing after doing the `os.listdir()` in the code.

Error message 
```
sr3 restart
stopping: sending SIGTERM ................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................... ( 515 ) Done
Waiting 1 sec. to check if 515 processes stopped (try: 0)
Traceback (most recent call last):
  File "/usr/bin/sr3", line 11, in <module>
	load_entry_point('metpx-sr3==3.2.0rc2', 'console_scripts', 'sr3')()
  File "/usr/lib/python3/dist-packages/sarracenia/sr.py", line 3447, in main
	gs.stop()
  File "/usr/lib/python3/dist-packages/sarracenia/sr.py", line 2495, in stop
	self._clean_missing_proc_state()
  File "/usr/lib/python3/dist-packages/sarracenia/sr.py", line 711, in _clean_missing_proc_state
	self._clean_missing_proc_state_dir(self.user_cache_dir)
  File "/usr/lib/python3/dist-packages/sarracenia/sr.py", line 695, in _clean_missing_proc_state_dir
	t = p.read_text().strip()
  File "/usr/lib/python3.6/pathlib.py", line 1196, in read_text
	with self.open(mode='r', encoding=encoding, errors=errors) as f:
  File "/usr/lib/python3.6/pathlib.py", line 1183, in open
	opener=self._opener)
  File "/usr/lib/python3.6/pathlib.py", line 1037, in _opener
	return self._accessor.open(self, flags, mode)
  File "/usr/lib/python3.6/pathlib.py", line 387, in wrapped
	return strfunc(str(pathobj), *args)
FileNotFoundError: [Errno 2] No such file or directory: 'sarra_get_copernicus_01.pid'
```

After the `sr3 restart` script fails, **all of the processes are in a `stop` state** ☹️ 
Doing a `sr3 start` afterwards resumes the processes to their normal operations.

I tried a bunch of ways to recreate the problem, all of which did not work.
1. Kill a process manually from its PID number, i.e. `kill $PID#` and doing a `sr3 restart` afterwards.
2. Remove the `*.pid` file from a process from `~/.cache/sr3/component/config/` and doing a `sr3 restart` afterwards.
3. Rebooting the server and doing a `sr3 restart` after it has come back up. (I only tried this, because this is the same scenario where the problem got triggered today).

Rob's PR, https://github.com/MetPX/sarracenia/pull/1595 seems to fix the issue by wrapping the affected cleanups by a try/except clause. However, I did notice something interesting. When the PID file throws a `FileNotFoundError`, the same log line always seems to be linked in the affected processes log file.

```
2026-03-20 19:16:06,454 13204 [ERROR] sarracenia.sr _clean_missing_proc_state_dir pid file sarra_config1_03.pid disappeared (race condition, see #1571), skipping cleanup

# less sarra_config1_03.log
2026-03-20 19:16:06,454 [INFO] sarracenia.flow close flow/close completed cleanly pid: 45527 sarra/config1 instance: 3

----------------------------

2026-03-23 13:16:05,040 40948 [ERROR] sarracenia.sr _clean_missing_proc_state_dir pid file sarra_config2_10.pid disappeared (race condition, see #1571), skipping cleanup

# less sarra_config2_10.log

2026-03-23 13:16:05,040 [INFO] sarracenia.flow close flow/close completed cleanly pid: 7717 sarra/get_dms-products-nswob instance: 10
```

Looking at that code execution, I don't see anything that would affect the PID file. I looked at the various `on_stop` calls just before this and also didn't find anything that would access the PID file.

--- a/sarracenia/sr.py
+++ b/sarracenia/sr.py
@@ -653,8 +653,7 @@ class sr_GlobalState:
                                             with p.open() as f:
                                                 t = f.read().strip()
                                     except FileNotFoundError:
-                                        logger.error("pid file %s disappeared (race condition, see #1571), treating as missing", filename)
-                                        missing.append([c, cfg, i])
+                                        logger.error("pid file %s disappeared (race condition, see #1571), skipping.", filename)
                                         continue
                                     if t.isdigit():
                                         pid = int(t)

--- a/sarracenia/sr.py
+++ b/sarracenia/sr.py
@@ -516,12 +516,16 @@ class sr_GlobalState:
                         for pathname in state_files:
                             p = pathlib.Path(pathname)
                             if p.suffix in ['.pid', '.qname', '.state', '.noVip']:
-                                if sys.version_info[0] > 3 or sys.version_info[
-                                        1] > 4:
-                                    t = p.read_text().strip()
-                                else:
-                                    with p.open() as f:
-                                        t = f.read().strip()
+                                try:
+                                    if sys.version_info[0] > 3 or sys.version_info[
+                                            1] > 4:
+                                        t = p.read_text().strip()
+                                    else:
+                                        with p.open() as f:
+                                            t = f.read().strip()
+                                except FileNotFoundError:
+                                    logger.error("state file %s disappeared (race condition, see #1571), skipping", pathname)
+                                    continue
                                 #print( 'read pathname:%s len: %d contents:%s' % ( pathname, len(t), t[0:10] ) )
                                 if len(t) == 0:
                                     continue
@@ -641,12 +645,17 @@ class sr_GlobalState:
                                 i_found.append(i)
                                 if i != 0:
                                     p = pathlib.Path(filename)
-                                    if sys.version_info[0] > 3 or sys.version_info[
-                                            1] > 4:
-                                        t = p.read_text().strip()
-                                    else:
-                                        with p.open() as f:
-                                            t = f.read().strip()
+                                    try:
+                                        if sys.version_info[0] > 3 or sys.version_info[
+                                                1] > 4:
+                                            t = p.read_text().strip()
+                                        else:
+                                            with p.open() as f:
+                                                t = f.read().strip()
+                                    except FileNotFoundError:
+                                        logger.error("pid file %s disappeared (race condition, see #1571), treating as missing", filename)
+                                        missing.append([c, cfg, i])
+                                        continue
                                     if t.isdigit():
                                         pid = int(t)
                                         if pid not in self.procs:
@@ -690,12 +699,16 @@ class sr_GlobalState:
                         for filename in os.listdir():
                             if filename[-4:] == '.pid':
                                 p = pathlib.Path(filename)
-                                if sys.version_info[0] > 3 or sys.version_info[
-                                        1] > 4:
-                                    t = p.read_text().strip()
-                                else:
-                                    with p.open() as f:
-                                        t = f.read().strip()
+                                try:
+                                    if sys.version_info[0] > 3 or sys.version_info[
+                                            1] > 4:
+                                        t = p.read_text().strip()
+                                    else:
+                                        with p.open() as f:
+                                            t = f.read().strip()
+                                except FileNotFoundError:
+                                    logger.error("pid file %s disappeared (race condition, see #1571), skipping cleanup", filename)
+                                    continue
                                 if t.isdigit():
                                     pid = int(t)
                                     if pid not in self.procs:

